### PR TITLE
fix TRAFODION-1555

### PR DESCRIPTION
**Pull request by nonstop-qfchen**
Please refer to JIRA1555 for resolution details.  Thanks

### CODE REVIEW

File: core/sql/optimizer/ScanOptimizer.cpp [hunk 1]
```diff
@@ -8590,6 +8595,7 @@ void MDAMCostWA::compute()
 // 	  // invalidate the cache
 // 	  disjunctsFR_.reset();
 // 	  disjunctsLR_.reset();
+          MDAM_DEBUG0(MTL2, "Mdam scan lost due to exceeding cost bound");
```
Comment (DaveBirdsall) [inline core/sql/optimizer/ScanOptimizer.cpp]: I LOVE this kind of debugging aid.

File: core/sql/optimizer/ScanOptimizer.cpp [hunk 2]
```diff
@@ -8703,11 +8713,14 @@ void MDAMCostWA::computeDisjunct()
   if(CURRSTMT_OPTDEFAULTS->indexEliminationLevel() != OptDefaults::MINIMUM
      && (!mdamForced_)
 	 && (CmpCommon::getDefault(RANGESPEC_TRANSFORMATION) == DF_ON )
+	 && (CmpCommon::getDefault(MDAM_APPLY_RESTRICTION_CHECK) == DF_ON )
```
Comment (DaveBirdsall) [inline core/sql/optimizer/ScanOptimizer.cpp]: The effect of this change is that we will pick MDAM more often? I take it the method "checkMDAMadditionalRestriction" is where the problem was?

**Comment by DaveBirdsall** [comment]
Looks good to me. Might be good for another pair of eyes to look at it.

File: core/sql/optimizer/ScanOptimizer.cpp (hunk 2 above)
Comment (nonstop-qfchen) [inline core/sql/optimizer/ScanOptimizer.cpp]: Yes, the costing is much better and we should trust it more :-).

On Fri, Oct 23, 2015 at 12:50 PM, DaveBirdsall <notifications@github.com>
wrote:

> In core/sql/optimizer/ScanOptimizer.cpp
> <https://github.com/apache/incubator-trafodion/pull/144#discussion_r42893895>
> :
>
> > @@ -8703,11 +8713,14 @@ void MDAMCostWA::computeDisjunct()
> >    if(CURRSTMT_OPTDEFAULTS->indexEliminationLevel() != OptDefaults::MINIMUM
> >       && (!mdamForced_)
> >  	 && (CmpCommon::getDefault(RANGESPEC_TRANSFORMATION) == DF_ON )
> > +	 && (CmpCommon::getDefault(MDAM_APPLY_RESTRICTION_CHECK) == DF_ON )
>
> The effect of this change is that we will pick MDAM more often? I take it
> the method "checkMDAMadditionalRestriction" is where the problem was?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/144/files#r42893895>.
>



-- 
Regards, --Qifan

File: core/sql/optimizer/ScanOptimizer.cpp (hunk 1 above)
Comment (nonstop-qfchen) [inline core/sql/optimizer/ScanOptimizer.cpp]: Yes, these debug messages can help with narrow down the problem quickly.

On Fri, Oct 23, 2015 at 12:48 PM, DaveBirdsall <notifications@github.com>
wrote:

> In core/sql/optimizer/ScanOptimizer.cpp
> <https://github.com/apache/incubator-trafodion/pull/144#discussion_r42893715>
> :
>
> > @@ -8590,6 +8595,7 @@ void MDAMCostWA::compute()
> >  // 	  // invalidate the cache
> >  // 	  disjunctsFR_.reset();
> >  // 	  disjunctsLR_.reset();
> > +          MDAM_DEBUG0(MTL2, "Mdam scan lost due to exceeding cost bound");
>
> I LOVE this kind of debugging aid.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/144/files#r42893715>.
>



-- 
Regards, --Qifan

**Comment by sandhyasun** [comment]
Looks good. Merging..

**Pull request merged**